lib/core/include/irods/procApiRequest.h

```diff
@@ -1,19 +1,86 @@
+/// \retval  <0 On failure.
 int procApiRequest(rcComm_t *conn,
                    int apiNumber,
                    const void *inputStruct,
                    const bytesBuf_t *inputBsBBuf,
                    void **outStruct,
                    bytesBuf_t *outBsBBuf);
 
+/// Executes an iRODS API request.
+///
+/// Grants the caller control over which packing instructions are used during serialization and
+/// deserialization of the data structures. This form of procApiRequest is useful for maintaining
+/// backward compatibility.
+///
+/// \warning This function is provided for handling changes in packing instructions for an API
+/// only. All other properties of the original API must be honored. The behavior of this function
+/// is undefined if preconditions are violated.
+///
+/// \param[in]  conn        The RcComm used for communication.
+/// \param[in]  apiNumber   The integer which identifies the API to execute. See apiNumberData.h.
+/// \param[in]  packingInstructionTable The table holding various packing instructions. All
+///                         packing instructions needed for serialization of the data structure
+///                         must be defined in the table. If passed \p NULL, the global packing
+///                         instruction table will be used.
+/// \param[in]  inputPackingInstruction The packing instruction to use for serialization. Cannot
+///                         be \p NULL.
+/// \param[in]  inputStruct The data to serialize, typically an input data structure. Can be
+///                         passed \p NULL if no input is necessary.
+/// \param[in]  inputBsBBuf The byte buffer to stream to the server. Can be passed \p NULL if no
+///                         input byte stream is necessary.
+/// \param[in]  outputPackingInstruction The packing instruction to use for deserialization.
+///                         Can be set to \p NULL if no output is expected.
+/// \param[out] outStruct   The output data structure to be filled by the server. The output will
+///                         be heap-allocated and assigned to the pointer. Can be passed \p NULL
+///                         if no output is expected.
+/// \param[out] outBsBBuf   The output byte buffer to be filled by the server. Can be passed
+///                         \p NULL if no output stream is expected.
+///
+/// \return An integer indicating success or failure.
+/// \retval >=0 On success.
+/// \retval  <0 On failure.
+///
+/// \since 4.3.5
+int procApiRequest_raw(rcComm_t* conn,
+                       int apiNumber,
+                       const struct PackingInstruction* packingInstructionTable,
+                       const char* inputPackingInstruction,
+                       const void* inputStruct,
+                       const bytesBuf_t* inputBsBBuf,
+                       const char* outputPackingInstruction,
+                       void** outStruct,
+                       bytesBuf_t* outBsBBuf);
```

Comment:
Is `procApiRequest_raw` the best name for this?

Reply:
How do we feel about this name?

I'm tempted to say it's likely fine. I don't see us ever changing existing packing instructions again. And looking at the other inputs, I don't see any additional room for customization - i.e. we'd just introduce an entirely new API at that point.

Thoughts?

Reply:
that sounds reasonable/correct.